Lib/contextlib.py

```diff
@@ -494,6 +497,127 @@ def _fix_exception_context(new_exc, old_exc):
         return received_exc and suppressed_exc
 
 
+# Inspired by discussions on http://bugs.python.org/issue13585
+class ExitStack(_BaseExitStack, AbstractContextManager):
+    """Context manager for dynamic management of a stack of exit callbacks
+    For example:
+        with ExitStack() as stack:
+            files = [stack.enter_context(open(fname)) for fname in filenames]
+            # All opened files will automatically be closed at the end of
+            # the with statement, even if attempts to open files later
+            # in the list raise an exception
+    """
+
+    def __enter__(self):
+        return self
+
+    def __exit__(self, *exc_details):
+        gen = self._shutdown_loop(*exc_details)
+        try:
+            is_sync, result = next(gen)
+            while 1:
+                assert is_sync
+                is_sync, result = gen.send(result)
+        except StopIteration as e:
+            return e.value
+
+    def close(self):
+        """Immediately unwind the context stack"""
+        self.__exit__(None, None, None)
+
+
+# Inspired by discussions on https://bugs.python.org/issue29302
+class AsyncExitStack(_BaseExitStack, AbstractAsyncContextManager):
+    """Async context manager for dynamic management of a stack of exit callbacks
+    For example:
+        async with AsyncExitStack() as stack:
+                  connections = [await stack.enter_async_context(get_connection()) for i in range(5)]
```

Comment:
Tabulation is off and many lines are longer than 79.  Please make sure that your PR follows PEP 8.